src/batch/src/executor2/sort_agg.rs

```diff
@@ -0,0 +1,635 @@
+            args: vec![Arg {
+                input: Some(InputRefExpr { column_idx: 0 }),
+                r#type: Some(ProstDataType {
+                    type_name: TypeName::Int32 as i32,
+                    ..Default::default()
+                }),
+            }],
+            return_type: Some(ProstDataType {
+                type_name: TypeName::Int64 as i32,
+                ..Default::default()
+            }),
+            distinct: false,
+        };
+
+        let sum_agg = AggStateFactory::new(&prost)?.create_agg_state()?;
+
+        let group_exprs: Vec<BoxedExpression> = vec![];
+        let agg_states = vec![sum_agg];
+        let fields = group_exprs
+            .iter()
+            .map(|e| e.return_type())
+            .chain(agg_states.iter().map(|e| e.return_type()))
+            .map(Field::unnamed)
+            .collect::<Vec<Field>>();
+        let executor = Box::new(SortAggExecutor2 {
+            agg_states,
+            group_keys: vec![],
+            sorted_groupers: vec![],
+            child: Box::new(child),
+            schema: Schema { fields },
+            identity: "SortAggExecutor".to_string(),
+            output_size_limit: DEFAULT_CHUNK_BUFFER_SIZE,
+        });
+
+        let mut stream = executor.execute();
+        let res = stream.next().await.unwrap();
+        assert_matches!(res, Ok(_));
+        assert_matches!(stream.next().await, None);
+
+        let actual = res?.column_at(0).array();
+        let actual: &I64Array = actual.as_ref().into();
+        let v = actual.iter().collect::<Vec<Option<i64>>>();
+        assert_eq!(v, vec![Some(6)]);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[allow(clippy::many_single_char_names)]
+    async fn execute_sum_int32_grouped() -> Result<()> {
+        use risingwave_common::array::ArrayImpl;
+        let anchor: Arc<ArrayImpl> = Arc::new(array_nonnull! { I32Array, [1, 2, 3, 4] }.into());
+
+        let chunk = DataChunk::builder()
+            .columns(vec![
+                Column::new(anchor.clone()),
+                Column::new(Arc::new(array_nonnull! { I32Array, [1, 1, 3, 3] }.into())),
+                Column::new(Arc::new(array_nonnull! { I32Array, [7, 8, 8, 9] }.into())),
+            ])
+            .build();
+
+        // mock a child executor
+        let schema = Schema {
+            fields: vec![
+                Field::unnamed(DataType::Int32),
+                Field::unnamed(DataType::Int32),
+                Field::unnamed(DataType::Int32),
+            ],
+        };
+        let mut child = MockExecutor::new(schema);
+        child.add(chunk);
+
+        let chunk = DataChunk::builder()
+            .columns(vec![
+                Column::new(anchor.clone()),
+                Column::new(Arc::new(array_nonnull! { I32Array, [3, 4, 4, 5] }.into())),
+                Column::new(Arc::new(array_nonnull! { I32Array, [9, 9, 9, 9] }.into())),
+            ])
+            .build();
+        child.add(chunk);
+
```

Comment:
Sorry for the slow review. I am still tracing the new `offset` & `limit` variables to understand how the chunks are reorganized into ones with proper size, and validating if there are missed edges cases.

Sharing one problem I just found:

With an extra input chunk here:
```
        let chunk = DataChunk::builder()
            .columns(vec![
                Column::new(anchor.clone()),
                Column::new(Arc::new(array_nonnull! { I32Array, [5, 5, 5, 5] }.into())),
                Column::new(Arc::new(array_nonnull! { I32Array, [9, 9, 9, 9] }.into())),
            ])
            .build();
        child.add(chunk);
```
The assertion on line 143 would fail.
This may be fixed by moving `processed_group_cnt = 0;` from line 199 to before `continue` on line 194.

Reply:
I am sorry. The maintenance of `processed_group_cnt` is confusing, and I have refined it to use a `left_capability` to cut the output chunk in proper size.